RecoTauTag/RecoTau/interface/TauWPThreshold.h

```diff
@@ -0,0 +1,64 @@
+#ifndef RecoTauTag_RecoTau_TauWPThreshold_h
+#define RecoTauTag_RecoTau_TauWPThreshold_h
+
+#include "DataFormats/TauReco/interface/BaseTau.h"
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include <TF1.h>
+
+namespace tau {
+  class TauWPThreshold {
+  public:
+    explicit TauWPThreshold(const std::string& cut_str) {
+      bool simple_value = false;
+      try {
+        size_t pos = 0;
+        value_ = std::stod(cut_str, &pos);
+        simple_value = (pos == cut_str.size());
+      } catch (std::invalid_argument&) {
+      } catch (std::out_of_range&) {
+      }
```

Comment:
This violates the [CMSSW coding rules](https://cms-sw.github.io/cms_coding_rules.html):
> 7.11 In general, do not catch exceptions – leave them to the Framework (see [Exception Guidelines](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEdmExceptionUse#Exception%20Handling)).

These changes should not have been accepted in the current state. Please fix this misuse ASAP.

Reply:
@mbluj can you have a look?

Reply:
OK

Reply:
Do I understand correctly that try-catch can be simply removed as exceptions will be cached by framework?
@fwyzard for my curiosity: those exception catches was found because the exceptions have been thrown or it was done with other means?

Reply:
> Do I understand correctly that try-catch can be simply removed as exceptions will be cached by framework?

Exceptions caught by the framework will usually cause the jobs to terminate.


> @fwyzard for my curiosity: those exception catches was found because the exceptions have been thrown or it was done with other means?

These exceptions have been noticed running `cmsRun hlt.py` under `gdb` and telling it to catch exceptions, in order to investigate were some failures are happening.
These `try`/`catch` blocks add _a lot_ of noise to the debugging sessions, making it harder to figure out were the real problems are happening.

In general, in CMSSW it is discouraged to use `try`/`catch` as a way to check if a condition is valid or not; it's better to do an explicit check.

This approach is idiomatic in python, where checking for an exception is almost as fast as checking for a return value. In C++, the regular return values are faster, and the exceptions are slower.

Reply:
Thanks. So it will be best to replace `try/catch` by an explicit condition check (not sure if simple here) or if not possible to throw `cms::Exception` instead of C++ one? Or leave it to FW as anyway an incorrect cut value will lead to either a crash in another place or an incorrect behavior?

Reply:
This was already discussed in https://github.com/cms-sw/cmssw/issues/39627 sometime ago and solutions without using try/catch were proposed. We just forgot to implement it during the code refactoring... sorry.

Reply:
@kandrosov if you have a fix already implemented please go with it (or provide to me). If not, I have prepare one probably based on `std::strtod`

Reply:
My proposal below:
```suggestion
      const char* cut_cstr = cut_str.c_str();
      char* end_cstr{};
      value_ = std::strtod(cut_cstr, &end_cstr);
      simple_value = (!*end_cstr);
      if (cut_cstr == end_cstr) {  // invalid cut_str
        throw cms::Exception("TauWPThreshold: invalid expression")
            << "Invalid expression of WP cut = '" << cut_str << "'.";
      }
      if (!std::isfinite(value_)) {
        throw cms::Exception("TauWPThreshold: not finite") << "WP threshold value is not finite, value = " << value_<< ".";
```
@kandrosov @fwyzard please let me know if it is fine, if so I will prepare a PR to master (and backports if needed).

Reply:
> @mbluj No, I planned to implement it for the code refactoring in this PR, but unfortunately, I forgot to do it...

OK, so I go on with my implementation if there are not objections. Further comments can anyway be made in a PR to be prepared.

Reply:
@mbluj having a parsing error at this stage is ok; it just means that WP is not a "simple value". So I think your code should be modified like this:
```c++
const char* cut_cstr = cut_str.c_str();
char* end_cstr;
value_ = std::strtod(cut_cstr, &end_cstr);
simple_value = end_cstr != cut_cstr && std::isfinite(value_);
```
(without any `throw`)

Reply:
Okay, the code I proposed above is also wrong. For example, for `cut_str="1 + 1"`, it will return `simple_value=true` and `_value=1`, which is not intended. So should be something like this:
```c++
    const char* cut_cstr = cut_str.c_str();
    char* end_cstr;
    value_ = std::strtod(cut_cstr, &end_cstr);
    const bool simple_value = end_cstr != cut_cstr && static_cast<std::size_t>(end_cstr - cut_cstr) == cut_str.size() && value_ != HUGE_VALF;
```

Reply:
it is enough to check if `end_cstr` is empty `!*end_cstr`, so the piece of code can be like this:
```
      const char* cut_cstr = cut_str.c_str();
      char* end_cstr{};
      value_ = std::strtod(cut_cstr, &end_cstr);
      simple_value = (!*end_cstr && cut_cstr != end_cstr && std::isfinite(value_));
      if (!simple_value) {
```
I think `std::isfinite(value_)` is better than `value_ != HUGE_VAL` as it does not accept `nan`. BTW, current code accepts both `nan` and `inf` as a valid "simple_value".

Reply:
Okay. Perhaps, as in your original proposal, an exception should be thrown for `inf` and especially `nan` - I don't see any use cases where one would want to use them in valid configs...
Is `std::infinite(HUGE_VALF) == false` guaranteed, or could it be compiler-dependent? To make sure that we account for all possible return scenarios.